### PR TITLE
gh-118527: Intern code consts in free-threaded build

### PR DESCRIPTION
**Pull request by colesbury**
We already intern and immortalize most string constants. In the free-threaded build, other constants are a frequent source of reference count contention because they are shared by all threads running the same code objects.

For example, the following function would not otherwise scale because `1.5` would be a non-immortalized constant shared by all threads that execute `run_in_parallel`.

```python
def run_in_parallel(x, iters):
    for _ in range(iters):
        x *= 1.5
    return x
```

* Issue: gh-118527

### CODE REVIEW

**Comment by colesbury** [comment]
!buildbot nogil

File: Objects/codeobject.c [hunk 1]
```diff
@@ -2399,3 +2452,183 @@ _PyCode_ConstantKey(PyObject *op)
...
+        if (PySet_GET_SIZE(op1) != PySet_GET_SIZE(op2)) {
+            return 0;
+        }
+        Py_ssize_t pos1 = 0, pos2 = 0;
+        PyObject *obj1, *obj2;
+        Py_hash_t hash1, hash2;
+        while ((_PySet_NextEntry(op1, &pos1, &obj1, &hash1)) &&
+               (_PySet_NextEntry(op2, &pos2, &obj2, &hash2)))
+        {
+            if (obj1 != obj2) {
```
Comment (DinoV) [inline Objects/codeobject.c:2522]: And seems like this should do a recursive `compare_constants` too

File: Objects/codeobject.c [hunk 2]
```diff
@@ -2399,3 +2452,183 @@ _PyCode_ConstantKey(PyObject *op)
...
+                return 0;
+            }
+        }
+        return 1;
+    }
+    else if (PySlice_Check(op1)) {
+        PySliceObject *s1 = (PySliceObject *)op1;
+        PySliceObject *s2 = (PySliceObject *)op2;
+        return (s1->start == s2->start &&
```
Comment (DinoV) [inline Objects/codeobject.c:2570]: Seems like this should be a recursive compare or `intern_constants` should also recursively intern slice's members?

File: Include/internal/pycore_code.h [hunk 3]
```diff
@@ -16,6 +18,14 @@ extern "C" {
 #define _PyCode_HAS_INSTRUMENTATION(CODE) \
     (CODE->_co_instrumentation_version > 0)
 
+struct _py_code_state {
+    PyMutex mutex;
```
Comment (DinoV) [inline Include/internal/pycore_code.h:22]: This should be in the `LOCKS_INIT` list in `pystate.c`

File: Objects/codeobject.c [hunk 4]
```diff
@@ -2399,3 +2452,183 @@ _PyCode_ConstantKey(PyObject *op)
...
+    PyObject *op = (PyObject *)key;
+    if (PySlice_Check(op)) {
+        PySliceObject *s = (PySliceObject *)op;
+        PyObject *data[3] = { s->start, s->stop, s->step };
+        return _Py_HashBytes(&data, sizeof(data));
+    }
+    else if (PyTuple_CheckExact(op)) {
+        Py_ssize_t size = PyTuple_GET_SIZE(op);
+        PyObject **data = _PyTuple_ITEMS(op);
+        return _Py_HashBytes(data, sizeof(PyObject *) * size);
```
Comment (colesbury) [inline Objects/codeobject.c:2605]: We first intern and de-dupe the contents so identity comparisons (including hasing pointers) is sufficient.

For example the following prints `True` with this PR:

```python
def foo():
    return (0.0, (1, 2, "hello"))

def bar():
    return (0.0, (1, 2, "hello"))

print(foo() is bar())
```

**Review by DinoV — APPROVED** [review]
Adding the mutex to `LOCKS_INIT` is really the most important thing, the slices are probably fine either way, it'd just be a little less duplication if we followed them, but they're probably usually in the small ints list anyway.

**Comment by colesbury** [comment]
!buildbot nogil

**Pull request merged**